src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs

```diff
@@ -0,0 +1,98 @@
+using Stratis.Features.FederatedPeg.Wallet;
+using Recipient = Stratis.Features.FederatedPeg.Wallet.Recipient;
+
+namespace Stratis.Features.FederatedPeg
+{
+    /// <summary>
+    /// Uses this node's federation wallet and the current network fee policy to build withdrawal transactions for the federation multisig.
+    /// </summary>
+    public class FedMultiSigWithdrawalTransactionBuilder
+    {
+        private readonly IFederationWalletManager federationWalletManager;
+
+        private readonly IFederatedPegSettings federatedPegSettings;
+        
+        private readonly Network network;
+
+        private readonly IWalletFeePolicy walletFeePolicy;
+
+        public FedMultiSigWithdrawalTransactionBuilder(
+            Network network,
+            IFederationWalletManager federationWalletManager,
+            IFederatedPegSettings federatedPegSettings,
+            IWalletFeePolicy walletFeePolicy)
+        {
+            this.network = network;
+            this.federationWalletManager = federationWalletManager;
+            this.federatedPegSettings = federatedPegSettings;
+            this.walletFeePolicy = walletFeePolicy;
+        }
+
+        /// <summary>
+        /// Builds a withdrawal transaction from the federation multisig to the given recipients.
+        /// Determines which coins from the federation wallet to spend for the total amount being sent to recipients.
+        /// Calculates the fee based on the wallet fee policy.
+        /// Sets the change address to be the multisig redeem script hash.
+        /// Signs the transaction using the given private keys.
+        /// </summary>
+        /// <param name="recipients">The recipients of the transaction.</param>
+        /// <param name="privateKeys">The private keys of the multisig.</param>
+        /// <returns></returns>
+        public Transaction BuildTransaction(List<Recipient> recipients, Key[] privateKeys)
+        {
+            // FederationWalletTransactionHandler only supports signing with a single key - the fed wallet key - so we don't use it to build the transaction.
+            // However we still want to use it to determine what coins we need, so hack this together here to pass in to FederationWalletTransactionHandler.DetermineCoins.
+            var multiSigContext = new Wallet.TransactionBuildContext(recipients);
+
+            (List<Coin> coins, List<Wallet.UnspentOutputReference> _) = FederationWalletTransactionHandler.DetermineCoins(this.federationWalletManager, this.network, multiSigContext, this.federatedPegSettings);
+
+            // MultiSigAddress from the wallet is not safe. It's only pulled from multisig-wallet.json and can
+            // be different from the *actual* multisig address for the current redeem script.
+            // Instead, use the address from settings - it's derived from the redeem script provided at startup.
+            var multiSigAddress = this.federatedPegSettings.MultiSigAddress.ScriptPubKey;
```

Comment:
What if they put the wallet in the data folder and start it up without changing their given script. Then they will have the reverse effect of yours?

Should we put in a check that the wallet matches the given redeemscript? Seems best, could be a different PR.

Reply:
That's a risk too. I suggested exactly what you did on the DevOps issue, think it's a separate PR.

Reply:
Since you have all the keys available you can make probably derive the redeem script within this code, although detecting if there's a difference between the wallet and settings is desirable as well I think.

Reply:
Agreed that it can be a separate PR though.